code/modules/mining/machinery/material_extractor.dm

```diff
@@ -0,0 +1,408 @@
+		else
+			to_chat(user, SPAN_NOTICE("Everything is working correctly."))
+
+	if(output_container)
+		var/output_desc = SPAN_NOTICE("It has a [output_container.name] in place to receive reagents.")
+		if(is_output_container_full())
+			output_desc = "[output_desc] [SPAN_WARNING("Its full!")]" 
+		to_chat(user, output_desc)
+	else 
+		to_chat(user, SPAN_NOTICE("It has nothing to pour reagents into."))
+
+/obj/machinery/atmospherics/unary/material/extractor/attackby(var/obj/item/I, var/mob/user)
+	if(istype(I, /obj/item/chems/glass))
+		add_fingerprint(user)
+		if(!output_container)
+			if(!user.unEquip(I, src))
+				return
+			output_container = I
+			user.visible_message(SPAN_NOTICE("\The [user] place \a [I] in \the [src]."), SPAN_NOTICE("You place \a [I] in \the [src]."))
+			update_icon()
+		return TRUE
+	return ..()
+
+/obj/machinery/atmospherics/unary/material/extractor/physical_attack_hand(var/mob/user)
+	if(output_container)
+		user.put_in_hands(output_container)
+		output_container.update_icon()
+		output_container = null
+		update_icon()
+		return TRUE
+	return ..()
+
+/obj/machinery/atmospherics/unary/material/extractor/power_change()
+	. = ..()
+	QUEUE_TEMPERATURE_ATOMS(src)
+
+/obj/machinery/atmospherics/unary/material/extractor/on_update_icon()
+	cut_overlays()
+	
+	var/initial_state = initial(icon_state)
+	if(!use_power || inoperable())
+		icon_state = "[initial_state]-off"
+	else 
+		icon_state = initial_state
+
+	if(panel_open)
+		add_overlay("[initial_state]-open")
+	if(output_container)
+		add_overlay("[initial_state]-bucket")
+
+/obj/machinery/atmospherics/unary/material/extractor/Process()
+	..()
+	if(!is_input_tank_empty())
+		update_use_power(POWER_USE_ACTIVE)
+
+		//Its crucial that all reactions happen ONLY in the input tank
+		if(!force_react_input())
+			process_input_tank()
+			input_clear_remainder() //Keep the input tank tidy by removing very small amounts of reagents
+	else
+		update_use_power(POWER_USE_IDLE)
+
+	//Make the internal liquid tank transfer liquids into the output
+	move_liquids_to_output()
+
+//For some reasons that's not in the unary base class...
+/obj/machinery/atmospherics/unary/material/extractor/return_air()
+	return air_contents
+
+//Remove trace amounts of reagents from the input tank, to prevent that from causing problems
+/obj/machinery/atmospherics/unary/material/extractor/proc/input_clear_remainder()
+	if(input_buffer.reagents.total_volume >= GAS_EXTRACTOR_MIN_REAGENT_AMOUNT)
+		return
+	input_buffer.reagents.clear_reagents()
+
+//Calculate the amount of reagents we can get from this. Returns a list with each materials and the amount of expected reagents
+/obj/machinery/atmospherics/unary/material/extractor/proc/gather_resulting_reagents_vol(var/obj/O)
+	var/list/processable = list()
+	for(var/k in O.matter)
+		if(can_process_material_name(k))
+			processable[k] = MATERIAL_UNITS_TO_REAGENTS_UNITS(O.matter[k])
```

Comment:
One issue with that is that this proc also strip out matter entries that can be processed from the item. Because if there's anything left that can't be processed in the ore, it gets spat out the back so the smelter can process it. And we don't want the ore to essentially possibly give double the yield that way.

Reply:
Behaviorally it should be identical to your current code, just without setting `processable` to a list for objects that have nothing in their matter list. It's a memory optimization rather than a logic change.

Reply:
I think you don't understand here. 
`#define LAZYSET(L, A, I) if(!L) { L = list(); } L[A] = I;`
Lazyset doesn't remove the element from the source list, which I need to do with line 205 that's missing from that snippet:
`O.matter -= k` 
I fixed it by adding a check if the matter list is empty before instancing a new list instead.

Reply:
Ah gotcha, my bad.